### PR TITLE
Add security injection test suite

### LINKED ISSUE #116: Tests: Add security injection test suite
## Summary

There are no tests for security injection attacks (SQL injection, Unicode attacks, symlink attacks, path traversal).

## Missing Test Coverage

1. **SQL injection** - Test token/artifact names with SQL metacharacters
2. **Unicode attacks** - Test path handling with Unicode normalization issues
3. **Symlink attacks** - Test that symlinks in artifact paths are rejected
4. **Path traversal** - Test `../` patterns in all path parameters
5. **Null byte injection** - Test paths with `\x00` characters

## Recommended Tests

```python
@pytest.mark.parametrize("malicious_input", [
    "../../../etc/passwd",
    "..\\..\\..\\windows\\system32",
    "foo\x00bar",
    "foo/../bar",
    "foo/./bar",
    "\u202e/path",  # Right-to-left override
])
def test_path_injection(malicious_input):
    # Test that malicious paths are rejected
    ...
```

## Location

- `tests/security/` - new test directory for security tests

## Reference

Found in comprehensive review 2026-01-12. See `docs/repo/magpie/review-2026-01-12/test-coverage-review.txt` for details.

---
Generated with Claude Code (Opus 4.5)

### PR DESCRIPTION
**Pull request by duplico**
## Summary
- Create `tests/security/` directory with comprehensive security tests
- Add tests for SQL injection, Unicode attacks, symlink attacks, path traversal, and null byte injection
- Tests verify security properties documented in `src/magpie/storage/paths.py`

## Test Coverage
The new test suite covers:
1. **SQL Injection** - Test token/artifact names with SQL metacharacters to verify no database errors are exposed
2. **Unicode Attacks** - Test path handling with Unicode normalization issues and homoglyphs
3. **Symlink Attacks** - Test that symlinks in artifact paths don't allow escaping storage root (currently xfail)
4. **Path Traversal** - Test `../` patterns and verify HTTP framework normalization
5. **Null Byte Injection** - Test paths with `\x00` characters are rejected in tag names
6. **Boundary Conditions** - Test empty paths, long paths, reserved segments, and hidden segments

## Test Results
```
47 passed, 1 xfailed
```

The xfail test documents that symlinks in the storage directory are currently followed, which should be reviewed for security implications.

## Test plan
- [x] Run `pytest tests/security/` - all tests pass
- [ ] CI pipeline should pass

Closes #116

(AI-generated via Claude Code)

### CODE REVIEW

**Comment by duplico** [comment]
## Addressing Review Comments

I've pushed commit e99efd7 to address all 14 review comments:

**Critical Bug Fixes:**
1. **Line 63 assertion logic** - Changed `or` to `and` so both conditions are properly checked
2. **Line 444 assertion** - Added explicit `assert` statement with proper variable extraction from response JSON

**Test Fixes:**
3. **Whitespace test** - Updated to document both behaviors: `normalize_artifact_path` preserves whitespace-only paths, but `validate_artifact_path` rejects them. Test now verifies both.
4. **Duplicate tests in TestPathTraversal** - Removed duplicate function-level tests. Added docstring note explaining that unit tests in `test_path_validation.py` cover these. Kept HTTP-level tests which provide unique integration value.
5. **Duplicate tests in TestBoundaryConditions** - Removed `test_empty_path_rejected` which duplicated unit tests. Kept HTTP-level integration tests for reserved/hidden segments.
6. **Unnecessary assignment (line 146)** - Removed redundant `files` variable that was redefined in the loop

**Fixture Fixes:**
7. **conftest.py cleanup** - Changed from `clear()` to individual `pop()` calls for proper cleanup that doesn't remove other overrides
8. **Fixture duplication** - The security conftest.py is intentionally separate from integration conftest.py to maintain test isolation. Both files define similar noop functions because security tests need explicit control over fixture lifecycle.

**Type Annotations:**
9-11. Added `StorageService` type annotation to all three `test_storage_service` parameters

**AI Disclosure:**
12-14. Added AI disclosure comments to docstrings in all three files

All 40 tests pass (39 passed, 1 xfailed as expected).

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+        self, client: TestClient, payload: str
+    ) -> None:
+        """SQL injection payloads should not expose database errors."""
+        response = client.get(f"/api/v1/artifacts/{payload}/latest/info")
+
+        # Should not return 200 (success) for these malicious paths
+        assert response.status_code != 200
+        # Should never expose database errors or SQL syntax
+        response_text = response.text.lower()
+        assert "sql" not in response_text or "sqlalchemy" not in response_text
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - changed `or` to `and` so both conditions are properly checked. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+        # The ../ part gets normalized by HTTP, leaving just the SQL part
+        payload = "../../../'; DROP TABLE artifacts; --"
+        response = client.post(f"/api/v1/upload/{payload}", files=files)
+
+        # Either normalized and stored (safe), or rejected
+        # The key is no SQL execution happens
+        assert response.status_code in (200, 400, 404)
+        if response.status_code == 200:
+            # SQL characters in path are just literal characters
+            assert "DROP" not in response.json().get("artifact_path", "DROP")
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - added explicit `assert` statement with proper variable extraction from response JSON. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+        """Whitespace-only path is treated as literal path segment.
+
+        Note: The normalize_artifact_path function only strips leading/trailing
+        slashes, not whitespace. A path of '   ' is technically valid (though
+        unusual). This is acceptable as whitespace in URLs gets encoded.
+        """
+        # Whitespace is preserved as a literal path segment
+        result = normalize_artifact_path("   ")
+        assert result == "   "  # Preserved as-is
+
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - updated test to document both behaviors: `normalize_artifact_path` preserves whitespace-only paths, but `validate_artifact_path` rejects them. Test now verifies both. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+traversal, and null byte injection.
+
+These tests verify that security validation in src/magpie/storage/paths.py
+and the server routes correctly handle potentially malicious input.
+
+Note: These tests focus on ensuring that malicious payloads cannot achieve
+their intended attack goals. The specific HTTP status code may vary (400, 401,
+404, 422, 500) depending on where the payload is rejected - the key is that
+the attack does not succeed.
+"""
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py:16]: Added AI disclosure comment to the module docstring: 'AI-assisted: Generated with Claude Code (Opus 4.5).' (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+            assert "reserved" in response.json().get("message", "").lower()
+
+    def test_hidden_segments_rejected(self, client: TestClient) -> None:
+        """Hidden path segments (starting with .) should be rejected."""
+        content = b"test content"
+        files = {"file": ("test.bin", io.BytesIO(content), "application/octet-stream")}
+
+        response = client.post("/api/v1/upload/test/.hidden/artifact", files=files)
+        assert response.status_code == 400
+        assert "cannot start with" in response.json().get("message", "").lower()
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - removed duplicate tests that directly call `normalize_artifact_path` and `validate_artifact_path`. These are already covered in `tests/unit/test_path_validation.py`. Kept the HTTP-level tests (`test_url_path_traversal_normalized_by_http` and `test_traversal_inside_path_normalized`) which provide unique integration-level value. Added docstring note explaining this. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+            ("//test//artifact//", "test/artifact"),  # Multiple slashes
+            ("test", "test"),  # Single segment
+        ],
+    )
+    def test_normalize_path_valid_inputs(self, path: str, expected: str) -> None:
+        """Valid paths should normalize correctly."""
+        result = normalize_artifact_path(path)
+        assert result == expected
+
+
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - removed `test_empty_path_rejected` which directly calls `normalize_artifact_path` (duplicate of unit tests). Kept `test_reserved_segments_rejected` and `test_hidden_segments_rejected` since they test via HTTP client and provide integration-level value. Added docstring note explaining the focus on HTTP-level tests. (AI-generated via Claude Code)

File: tests/security/conftest.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+"""Shared fixtures for security tests."""
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py]: Added AI disclosure comment to the docstring: 'AI-assisted: Generated with Claude Code (Opus 4.5).' (AI-generated via Claude Code)

File: tests/security/conftest.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    def override_storage_service() -> StorageService:
+        return test_storage_service
+
+    app.dependency_overrides[get_storage_service] = override_storage_service
+    app.dependency_overrides[require_admin_scope] = _noop_require_admin_scope
+    app.dependency_overrides[require_admin_scope_header] = _noop_require_admin_scope_header
+    app.dependency_overrides[require_write_scope] = _noop_require_write_scope
+    yield TestClient(app)
+    app.dependency_overrides.clear()
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py]: The security conftest.py intentionally defines its own fixtures for test isolation - security tests should not depend on or be affected by changes in integration test fixtures. The fixture duplication is deliberate for this reason. (AI-generated via Claude Code)

[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py]: Fixed - changed from `clear()` to individual `pop()` calls for each dependency override. This properly removes only the overrides set by this fixture without affecting any other overrides that may have been set. (AI-generated via Claude Code)

File: tests/security/__init__.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+"""Security test suite for magpie.
+
+This package contains tests for various security attack vectors including:
+- SQL injection via token/artifact names
+- Unicode normalization attacks on paths
+- Symlink attacks in artifact paths
+- Path traversal attacks with ../ patterns
+- Null byte injection in paths
+"""
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/__init__.py:11]: Added AI disclosure comment to the docstring: 'AI-assisted: Generated with Claude Code (Opus 4.5).' (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+    These tests verify that Unicode characters that look similar to ASCII
+    or have special normalization properties cannot be used to bypass
+    path validation or cause confusion.
+    """
+
+    def test_unicode_path_traversal_blocked(self, client: TestClient) -> None:
+        """Unicode lookalike dots should not allow path traversal."""
+        content = b"test content"
+        files = {"file": ("test.bin", io.BytesIO(content), "application/octet-stream")}
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - removed the unnecessary `files` variable assignment on line 146. The variable is redefined inside the loop on line 155 before use. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+    SQL_INJECTION_TAG_PAYLOADS = [
+        "'; DROP TABLE tags; --",
+        "test' OR '1'='1",
+        "latest; DELETE FROM *",
+        'v1.0"; DROP TABLE --',
+    ]
+
+    @pytest.mark.parametrize("payload", SQL_INJECTION_TAG_PAYLOADS)
+    def test_sql_injection_in_tag_name_create(
+        self, client: TestClient, test_storage_service, payload: str
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - added `StorageService` type annotation from `magpie.storage.service`. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 12]
```diff
@@ -0,0 +1,504 @@
...
+
+    Note: Currently symlinks in the storage directory are followed, which
+    could be a security concern. This test documents the current behavior.
+    """
+
+    @pytest.mark.xfail(
+        reason="Symlink following is currently allowed - see issue for security review"
+    )
+    def test_symlink_in_artifact_path_rejected(
+        self, client: TestClient, test_storage_service, tmp_path: Path
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - added `StorageService` type annotation from `magpie.storage.service`. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 13]
```diff
@@ -0,0 +1,504 @@
...
+    Null bytes can truncate strings in C-based systems and potentially
+    bypass extension checks or path validation. These tests verify that
+    null bytes are properly handled.
+
+    Note: Python/FastAPI may handle null bytes differently than C-based
+    systems. The key security property is that null bytes cannot be used
+    to bypass security controls or access unintended files.
+    """
+
+    def test_null_byte_in_tag_name(self, client: TestClient, test_storage_service) -> None:
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - added `StorageService` type annotation from `magpie.storage.service`. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+        response = client.post(f"/api/v1/upload/{payload}", files=files)
+
+        # Either normalized and stored (safe), or rejected
+        # The key is no SQL execution happens
+        assert response.status_code in (200, 400, 404)
+        if response.status_code == 200:
+            # SQL characters in path are just literal characters
+            data = response.json()
+            artifact_path = data.get("artifact_path", "")
+            assert "DROP" not in artifact_path
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed in commit 7bf7b2a. Updated the assertion to verify that SQL injection characters ARE safely stored as literal path characters (because file-based storage treats them as harmless plain text). The test now correctly verifies the security property: that SQL characters cannot execute, they are just stored as-is.

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 15]
```diff
@@ -0,0 +1,487 @@
...
+    @pytest.mark.parametrize("payload", SQL_INJECTION_TAG_PAYLOADS)
+    def test_sql_injection_in_tag_flush(self, client: TestClient, payload: str) -> None:
+        """SQL injection in flush tag endpoint should be rejected."""
+        response = client.post(
+            f"/api/v1/tags/{payload}/flush",
+            params={"confirm_walk_filesystem": True},
+        )
+
+        # Should be rejected by path validation (422 from regex) or 401 (auth)
+        assert response.status_code in (401, 422)
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed in commit 7bf7b2a. Removed the '401 (auth)' mention since auth dependencies are overridden to no-op in the test fixtures. Only 422 validation errors from the path regex can occur.

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+            "..\x00/etc/passwd",
+            "test/../\x00../../etc/passwd",
+        ]
+
+        for payload in traversal_with_null:
+            # If the null byte truncates before .., it might pass
+            # But if .. is present, it should be caught
+            if ".." in payload.split("\x00")[0]:
+                with pytest.raises(InvalidArtifactPathError):
+                    normalize_artifact_path(payload.split("\x00")[0])
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed in commit 7bf7b2a. Rewrote the test to properly verify null byte handling:

1. **HTTP layer defense**: Added test confirming that httpx.URL rejects null bytes with InvalidURL exception - this means null bytes cannot reach the application via HTTP at all (defense-in-depth)

2. **normalize_artifact_path behavior**: Now directly tests that the function processes full strings including null bytes without C-style truncation. Payloads with '..' before or after null bytes are correctly rejected.

3. **No truncation verification**: Added assertion that paths with null bytes but no traversal are processed in full (null byte is just another character in Python strings)

(AI-generated via Claude Code)

File: tests/security/test_injection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Additional fix in commit 835aee0: Split the compound assertion into two separate `assert` statements for clearer failure messages when debugging test failures. (AI-generated via Claude Code)

File: tests/security/conftest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py]: Additional fix in commit 835aee0: Added detailed docstring comment explaining that security tests deliberately define their own auth dependency overrides for explicit isolation from integration tests. Also removed unnecessary explicit `return None` statements from the noop functions since Python implicitly returns None. (AI-generated via Claude Code)

File: tests/security/conftest.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      focus on input validation, not authentication)
+    """
+
+    def override_storage_service() -> StorageService:
+        return test_storage_service
+
+    app.dependency_overrides[get_storage_service] = override_storage_service
+    app.dependency_overrides[require_admin_scope] = _noop_require_admin_scope
+    app.dependency_overrides[require_admin_scope_header] = _noop_require_admin_scope_header
+    app.dependency_overrides[require_write_scope] = _noop_require_write_scope
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py:86]: Fixed - the fixture now saves and restores the previous state of dependency overrides rather than just popping them. This defensive approach handles the case where other test modules (like integration tests) might have set overrides that run before security tests in a combined test run.

The fixture now:
1. Saves existing override values before setting its own
2. After the test, restores the saved values (or removes if there was none)

This ensures complete isolation regardless of pytest discovery order or whether tests are run individually or together.

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+            "/api/v1/upload/escape_link/artifact",
+            files=files,
+        )
+
+        # The storage service should handle this safely
+        # Either reject or store in the actual symlink location (not follow it)
+        if response.status_code == 200:
+            # If upload succeeded, verify it didn't write to the target
+            target_file = target_path / "artifact"
+            assert not (target_file / "blobs").exists(), (
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - renamed `target_file` to `target_artifact_dir` to accurately reflect that it represents a directory path, not a file.

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+        # Should return 400 (rejected) or 200 with empty results
+        assert response.status_code in (200, 400)
+        if response.status_code == 200:
+            # If accepted, should return safe empty result
+            data = response.json()
+            assert "paths" in data
+            # SQL injection shouldn't return data from other tables
+            for path in data["paths"]:
+                assert "DROP" not in path
+                assert "SELECT" not in path
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - replaced the string-based checks for "DROP" and "SELECT" with a more meaningful assertion: verifying that any returned paths start with the prefix filter.

Since magpie uses file-based storage (not SQL), the prefix parameter is used for literal string matching. If SQL injection were somehow possible, results from "other tables" wouldn't match the prefix. By asserting `path.startswith(payload)`, we verify the file-based filtering works correctly and no SQL execution occurred.

This approach:
1. Eliminates false positives from legitimate paths containing "DROP" or "SELECT"
2. Tests the actual security property (file-based prefix filtering works)
3. Would detect actual SQL injection (results wouldn't match the prefix)

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 20]
```diff
@@ -0,0 +1,514 @@
...
+            "\u2025/etc/passwd",  # Two dot leader
+        ]
+
+        for payload in unicode_traversal_payloads:
+            files = {"file": ("test.bin", io.BytesIO(content), "application/octet-stream")}
+            response = client.post(f"/api/v1/upload/{payload}", files=files)
+            # The path traversal component should be blocked
+            assert response.status_code != 200 or ".." not in response.json().get(
+                "artifact_path", ""
+            )
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed - the assertion now verifies the actual security property: that the stored artifact path is within the storage root directory.

The original assertion checked for ASCII ".." which doesn't apply when using Unicode lookalikes like U+2024 (ONE DOT LEADER) and U+2025 (TWO DOT LEADER). The new implementation:

1. Gets the storage_root from the test_storage_service fixture
2. If the upload succeeds (status 200), resolves the full filesystem path
3. Asserts that the resolved path is relative to (i.e., within) the storage root using `Path.is_relative_to()`

This directly tests the security property we care about: Unicode lookalikes cannot be used to escape the storage directory, regardless of how they're handled internally.

(AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 21]
```diff
@@ -0,0 +1,533 @@
...
+        rejected by validate_artifact_path as an invalid artifact path.
+        """
+        # Whitespace is preserved as a literal path segment by normalize_artifact_path
+        normalized = normalize_artifact_path("   ")
+        assert normalized == "   "
+
+        # But validation should reject whitespace-only paths
+        with pytest.raises(InvalidArtifactPathError):
+            validate_artifact_path(normalized)
+
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Removed. The unit test test_reject_whitespace_only already covers this validation behavior. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 22]
```diff
@@ -0,0 +1,533 @@
...
+    Note: Core path normalization and traversal rejection behavior for
+    normalize_artifact_path / validate_artifact_path is covered in depth in
+    tests/unit/test_path_validation.py::TestNormalizeArtifactPath.
+    To avoid duplicated coverage and maintenance burden, this security test
+    suite focuses on HTTP-level traversal and encoding behaviors rather than
+    re-testing the same normalization inputs here.
+
+    Plain '../' in URLs is normalized by the HTTP framework before
+    reaching the application, so these tests verify HTTP-level behavior.
+    """
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py:310]: Addressed. Updated the TestPathTraversal docstring to clarify the value of integration-level security tests: unit tests verify individual validation functions work correctly, while security tests verify the complete HTTP request pipeline handles attacks correctly - including HTTP framework URL normalization, route parsing, and response handling. Both layers matter: unit tests catch bugs in validation logic, while integration tests catch misconfigurations (e.g., a route that forgets to call validation). (AI-generated via Claude Code)

File: tests/security/conftest.py [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+    This provides explicit isolation - security tests run in their own test
+    directory and should not depend on fixtures from other test modules. The
+    duplication is intentional for clarity and to avoid cross-module coupling.
+
+    Unlike integration tests which use an autouse fixture to override auth
+    dependencies globally, this fixture explicitly manages the dependency
+    lifecycle per-test. This prevents any potential conflict since:
+    1. Security tests are in a separate directory (tests/security/)
+    2. pytest conftest.py fixtures are scoped to their directory
+    3. Each fixture saves and restores state independently
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/conftest.py]: Addressed. Simplified the comment to acknowledge the similarity to integration tests while noting they may evolve independently. Referenced issue #113 which tracks consolidating shared fixtures across the test suite. (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 24]
```diff
@@ -0,0 +1,533 @@
...
+    These tests verify that the system properly handles or rejects
+    symlinks that could be used to escape the storage directory or
+    access unauthorized files.
+
+    Note: Currently symlinks in the storage directory are followed, which
+    could be a security concern. This test documents the current behavior.
+    """
+
+    @pytest.mark.xfail(
+        reason="Symlink following is currently allowed - see issue for security review"
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Fixed. Created issue #135 (Security: Review symlink following behavior in storage) and updated the xfail reason to reference it: "Symlink following is currently allowed - see issue #135 for security review". (AI-generated via Claude Code)

File: tests/security/test_injection.py [hunk 25]
```diff
@@ -0,0 +1,533 @@
...
+            assert "reserved" in response.json().get("message", "").lower()
+
+    def test_hidden_segments_rejected(self, client: TestClient) -> None:
+        """Hidden path segments (starting with .) should be rejected."""
+        content = b"test content"
+        files = {"file": ("test.bin", io.BytesIO(content), "application/octet-stream")}
+
+        response = client.post("/api/v1/upload/test/.hidden/artifact", files=files)
+        assert response.status_code == 400
+        assert "cannot start with" in response.json().get("message", "").lower()
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Removed. The reserved segment tests (blobs, metadata, .magpie) are already covered by unit tests in test_path_validation.py (test_reject_reserved_blobs, test_reject_reserved_metadata, test_reject_reserved_magpie, test_reserved_at_start, test_reserved_in_middle). Since these HTTP-level tests were just calling the same validation logic without adding unique integration-level value, they've been removed. (AI-generated via Claude Code)

[the remark this answers is not in the document]
Comment (duplico) [reply tests/security/test_injection.py]: Removed. The hidden segment test is already covered by unit test test_reject_hidden_segment in test_path_validation.py. Removed to avoid duplicate coverage. (AI-generated via Claude Code)
